File: 수학_확률 및 통계/leetcode382(Medium)-Linked List Random Node.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 */
var Solution = function (head) {
  this.head = head
}

/**
 * @return {number}
 */
Solution.prototype.getRandom = function () {
  let result = null
  let i = 1

  let node = this.head
  while (node) {
    if (Math.random() < 1 / i) {
      result = node.val
    }
    node = node.next
    i += 1
  }
  return result
}

File: 수학_확률과 통계/leetcode382(Medium)-Linked List Random Node.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# Reservoir sampling (저수지 샘플링) 알고리즘 : 시간복잡도 O(n), 공간복잡도 O(1)
class Solution:

    def __init__(self, head: Optional[ListNode]):
        self.head = head

    def getRandom(self) -> int:
        result = None
        i = 1

        node = self.head
        while node:
            if random.random() < (1 / i): # Reservoir sampling
                result = node.val
            node = node.next
            i += 1
        return result